flocker/node/agents/ebs.py

```diff
@@ -682,13 +786,38 @@ def _next_device(self, instance_id, volumes, devices_in_use):
 
     def create_volume(self, dataset_id, size):
         """
+        Create a volume on EBS backend.
+        """
+        return self.create_volume_with_profile(
+            dataset_id, size, MandatoryProfiles.DEFAULT.value)
+
+    def create_volume_with_profile(self, dataset_id, size, profile_name):
+        """
         Create a volume on EBS. Store Flocker-specific
         {metadata version, cluster id, dataset id} for the volume
         as volume tag data.
         Open issues: https://clusterhq.atlassian.net/browse/FLOC-1792
         """
-        requested_volume = self.connection.create_volume(
-            size=int(Byte(size).to_GiB().value), zone=self.zone)
+        requested_size = int(Byte(size).to_GiB().value)
+        try:
+            volume_type, iops = _volume_type_and_iops_for_profile_name(
+                profile_name, requested_size)
+            requested_volume = self.connection.create_volume(
+                size=requested_size,
+                zone=self.zone,
+                volume_type=volume_type,
+                iops=iops)
+        except EC2ResponseError as e:
+            CREATE_VOLUME_FAILURE(dataset_id=dataset_id,
+                                  aws_code=e.code,
+                                  aws_message=e.message).write()
+            volume_type, iops = _volume_type_and_iops_for_profile_name(
```

Comment:
> Why are you retrying? Reading tests suggests this is here to cover cases where size doesn't match profile requirements. 

Sorry for the poor naming of ``test_create_too_small_volume_gold_profile``. Retrying in case volume creation failed because profile attributes corresponding to the requested profile cannot be satisfied for the volume that is being created.

As an example, ``test_create_too_small_volume_gold_profile`` attempts to create volume with:
size: 1 GiB
profile (``GOLD``):
volume_type: ``io1``
iops:size ratio: 30:1 (thats the best iops ratio for ``io1`` volume type (not documented, but noticed during testing of hack day project)).

With these parameters, EBS volume creation fails with the following error:

```
<Response><Errors><Error><Code>InvalidParameterValue</Code><Message>Volume iops of 30 is too low; minimum is 100.</Message></Error></Errors><RequestID>3f9facc0-7e77-4b7c-9655-69e6db504f7f</RequestID></Response>
```

The retry is saying that we have tried our best to honor profile request during volume creation. If we failed, we will make one more attempt to create your volume with default profile. Profile assignment can be a best effort attempt. If we create the profile and move on, profile compliance module (to be built later on) can detect and act on profile violations at run time (similar to https://pubs.vmware.com/vsphere-51/index.jsp?topic=%2Fcom.vmware.vsphere.storage.doc%2FGUID-46345561-C7ED-4296-91AE-8F6C5D8FF552.html).

Apologies for not documenting the reason for retry. Will add it.